pregenerate_exercises.py

```diff
@@ -0,0 +1,284 @@
+            "If the RST file includes multiple submit-directives, you will be asked to confirm personalized "
+            "instance generation for each one. By default, all personalized exercises in the course are generated.")
+    # Optional arguments
+    parser.add_argument("--instances", type=int, default=10, dest="instances",
+                        help="Number of instances to generate for an exercise")
+    parser.add_argument("--keep-old", action="store_true", dest="keep_old", default=False,
+                        help="Keep existing generated instances instead of deleting them first")
+    parser.add_argument("--gen-if-none-exist", action="store_true", dest="gen_if_none_exist", default=False,
+                        help="Only generate new instances if no instances exist yet")
+
+    args = parser.parse_args()
+
+    exercise_rst = args.exercise_rst
+
+    all_rst_files = []
+    if exercise_rst:
+        all_rst_files.append(exercise_rst)
+    else:
+        for root, dirs, files in os.walk('.'):
+            for filename in files:
+                if filename.endswith(".rst"):
+                    all_rst_files.append(os.path.join(root, filename))
+    if not all_rst_files:
+        print("The course has no personalized exercises so no instances are generated.")
+        exit()
+
+    all_keys_and_exercises = []
+    for rst_file in all_rst_files:
+        rst_data = parse_rst(rst_file)
+        for data in rst_data:
+            exercise_config = data['config']
+            exercise_key = data['key']
+            exercise = parse_config(exercise_config)
+            if exercise is None:
+                raise Exception("Exercise config not found: %s" % exercise_config)
+            if "personalized" in exercise and exercise["personalized"]:
+                exercise['key'] = '_'.join(
+                    part for part in os.path.splitext(rst_file)[0].split('/') if part not in ['.', '..']
+                ) + '_' + exercise_key
+                all_keys_and_exercises.append({'key': exercise_key, 'exercise': exercise})
+
+    print("Found personalized exercises with the following keys:")
+    for data in all_keys_and_exercises:
+        print(data['key'])
+    print()
+
+    exercises = []
+    confirm_all = False
+    for data in all_keys_and_exercises:
+        confirm_string = None
+        if not confirm_all:
+            confirm_string = input(
+                f"Include personalized exercise with the key '{data['key']}'? "
+                "Yes(Y) / No(N) / Yes to all(A) / Cancel(C) "
+            ).strip().lower()
+        if confirm_string == 'a':
+            confirm_all = True
+        if confirm_all or confirm_string == 'y':
+            exercises.append(data['exercise'])
+        elif confirm_string == 'c':
+            print("Cancelled.")
+            exit()
+        elif confirm_string != 'n':
+            raise Exception(f"Invalid input '{confirm_string}'")
```

Comment:
Raising exception may not be the best option here. If user mistypes after accepting 20 earlier items, and the script crashes and needs to be rerun, it will be quite annoying. I would rather ask the user to retype the command.